v3/src/components/graph/scatterdots.tsx

```diff
@@ -11,159 +12,135 @@ export const ScatterDots = memo(function ScatterDots(props: {
-        if (datum.selected) {
-          selectedDataObjects.current[datum.id] = {x: datum.x, y: datum.y}
-        }
-      })
 
-      target.current.classed('dragging', true)
-    }, [setData, data, setHighlightCounter, firstTime, setFirstTime]),
+        worldDataRef.current?.selectCases([tItsID])
+        // Record the current values so we can change them during the drag and restore them when done
+        worldDataRef.current?.selection.forEach(anID => {
+          selectedDataObjects.current[anID] = {
+            x: worldDataRef.current?.getNumeric(anID, xAttrID) ?? 0,
+            y: worldDataRef.current?.getNumeric(anID, yAttrID) ?? 0
+          }
+        })
+      }
+    }, [firstTime, setFirstTime, xAttrID, yAttrID, worldDataRef]),
 
     onDrag = useCallback((event: MouseEvent) => {
-      if (dragID >= 0) {
+      if (dragID !== '') {
         const newPos = {x: event.clientX, y: event.clientY},
           dx = newPos.x - currPos.current.x,
           dy = newPos.y - currPos.current.y
         currPos.current = newPos
         if (dx !== 0 || dy !== 0) {
           const deltaX = Number(xScale?.invert(dx)) - Number(xScale?.invert(0)),
             deltaY = Number(yScale?.invert(dy)) - Number(yScale?.invert(0))
-          setData((prevData) => {
-              prevData.forEach(datum => {
-                if (datum.selected) {
-                  datum.x += deltaX
-                  datum.y += deltaY
-                }
-              })
-              return prevData
-            }
-          )
+          worldDataRef.current?.selection.forEach(anID => {
+            const currX = worldDataRef.current?.getNumeric(anID, xAttrID),
+              currY = worldDataRef.current?.getNumeric(anID, yAttrID)
+            worldDataRef.current?.setValue(anID, xAttrID, Number(currX ?? 0) + deltaX)
+            worldDataRef.current?.setValue(anID, yAttrID, Number(currY ?? 0) + deltaY)
+          })
           setRefreshCounter(prevCounter => ++prevCounter)
         }
       }
-    }, [setData, dragID, xScale, yScale, setRefreshCounter]),
+    }, [dragID, xScale, yScale, setRefreshCounter, xAttrID, yAttrID, worldDataRef]),
 
     onDragEnd = useCallback(() => {
-      if (dragID >= 0) {
+      if (dragID !== '') {
         target.current
           .classed('dragging', false)
           .transition()
           .attr('r', defaultRadius)
-        setDragID(() => -1)
+        setDragID(() => '')
         target.current = null
 
-        setData(prevData => {
-          prevData.forEach(datum => {
-            const sDatum = selectedDataObjects.current[datum.id]
-            if (sDatum) {
-              datum.x = sDatum.x
-              datum.y = sDatum.y
-            }
-          })
-          return prevData
+        worldDataRef.current?.selection.forEach(anID => {
+          worldDataRef.current?.setValue(anID, xAttrID, selectedDataObjects.current[anID].x)
+          worldDataRef.current?.setValue(anID, yAttrID, selectedDataObjects.current[anID].y)
```

Comment:
As described in a comment above, this could be replaced with a single call to `setCaseValues()`.